### PR TITLE
Refactor QubitOperator expectation value (part 1)

### LINKED ISSUE #667: VQEProblem on 9 qubits getting killed without error message
### Priority

Medium

### Responsible

_No response_

### Qrisp Version

0.8.2

### Python Version

3.11.13

### Operating System

Linux

### Short Description

My VQEProblem with a 9-qubit diagonal Hamiltonian cannot run, but gets killed without further error message. Edit: Jax/Pennylane clash could be resolved (Pennylane update missing), so that's not the cause. VQEProblem is still killed though.

### Steps to Reproduce

1. Run the below example (VQEProblem on 9 qubits, diagonal Hamiltonian).

```python
import numpy as np
from qrisp import QuantumVariable
from qrisp.vqe import VQEProblem
from qrisp.operators.qubit import Z
import sys

# Hamiltonian
H = (
    -160.325456118848 * Z(0)
    + 50.0 * Z(0) * Z(1)
    + 50.0 * Z(0) * Z(2)
    + 50.0 * Z(0) * Z(3)
    + 21.2266047150761 * Z(0) * Z(4)
    + 50.0 * Z(0) * Z(6)
    + 14.2222483680606 * Z(0) * Z(7)
    - 170.897706166273 * Z(1)
    + 50.0 * Z(1) * Z(2)
    + 21.2266047150761 * Z(1) * Z(3)
    + 50.0 * Z(1) * Z(4)
    + 21.2266047150761 * Z(1) * Z(5)
    + 14.2222483680606 * Z(1) * Z(6)
    + 50.0 * Z(1) * Z(7)
    + 14.2222483680606 * Z(1) * Z(8)
    - 160.325456118848 * Z(2)
    + 21.2266047150761 * Z(2) * Z(4)
    + 50.0 * Z(2) * Z(5)
    + 14.2222483680606 * Z(2) * Z(7)
    + 50.0 * Z(2) * Z(8)
    - 166.293549415572 * Z(3)
    + 50.0 * Z(3) * Z(4)
    + 50.0 * Z(3) * Z(5)
    + 50.0 * Z(3) * Z(6)
    + 16.0642145527729 * Z(3) * Z(7)
    - 174.581638535698 * Z(4)
    + 50.0 * Z(4) * Z(5)
    + 16.0642145527729 * Z(4) * Z(6)
    + 50.0 * Z(4) * Z(7)
    + 16.0642145527729 * Z(4) * Z(8)
    - 166.293549415572 * Z(5)
    + 16.0642145527729 * Z(5) * Z(7)
    + 50.0 * Z(5) * Z(8)
    - 135.298326507515 * Z(6)
    + 50.0 * Z(6) * Z(7)
    + 50.0 * Z(6) * Z(8)
    - 160.572925841667 * Z(7)
    + 50.0 * Z(7) * Z(8)
    - 135.298326507515 * Z(8)
)


def hea_ry_ansatz(qv, params):
    from qrisp import ry

    for i in range(qv.size):
        ry(params[i], qv[i])

qarg = QuantumVariable(9)
problem = VQEProblem(hamiltonian=H, ansatz_function=hea_ry_ansatz, num_params=9)
energy = problem.run(qarg, depth=1, max_iter=3)
print(f"SUCCESS. Energy: {energy}", flush=True)
```

### Expected Behavior

Either:
- VQEProblem completing the run, outputting some energy
Or:
- error message detailing the issue (memory overflow?)

### Actual Behavior

Process gets killed a few seconds after starting VQEProblem.run(). A RuntimeWarning detailing the dependency clash gets printed. QAOAProblem on the same problem runs without problems, so I suspect the error is linked to the expectation value estimation of the VQE operator.

Output:
WARNING:2026-06-19 11:45:13,384:jax._src.xla_bridge:864: An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
/users/benedikt.poggel/miniforge3/envs/decisiontree-dev/lib/python3.11/site-packages/pennylane/__init__.py:196: RuntimeWarning: PennyLane is not yet compatible with JAX versions > 0.6.2. You have version 0.7.1 installed. Please downgrade JAX to 0.6.2 to avoid runtime errors using python -m pip install jax~=0.6.0 jaxlib~=0.6.0
  warnings.warn(
Killed

### Means of Verification

| Test-ID | Type     | Description | Expected Result |
|---------|----------|-------------|-----------------|
| T-001   | Positive |             |                 |
| T-002   | Negative |             |                 |
| T-003   | Boundary |             |                 |
| T-004   | Boundary |             |                 |

### Additional Context

_No response_

### Definition of Done

- [ ] Root cause identified
- [ ] Fix implemented and peer-reviewed
- [ ] Regression test added
- [ ] All tests pass (CI green)
- [ ] Changelog entry added

### PR DESCRIPTION
**Pull request by renezander90**
## Description

This PR refactors `QubitOperator.expectation_value` measurement planning and adds explicit shot-budget controls.

The existing `precision` parameter is unchanged and remains the default allocation mode. When `shots` is omitted, the estimated shot count scales approximately as `1 / precision**2` and with the Hamiltonian coefficient magnitudes through operator-derived measurement variances. These are state-independent estimates, so the resulting shot count and achieved precision are not guarantees.

The new public default `max_shots=100_000` prevents unexpectedly large precision-based measurements. Passing `max_shots=None` restores the previous unlimited-shot behavior.

## Related Issues

Closes #667, #649 
Related to #

## Type of Change

- [x] Feature (new functionality)
- [x] Change Request (modification of existing functionality)
- [ ] Bug Fix
- [x] Refactoring (no behavior change)
- [ ] Performance improvement
- [x] Documentation
- [ ] CI / Build

## Breaking Change?

- [x] Yes
- [ ] No

The default behavior now raises `ValueError` when the estimated precision-based shot count exceeds `100_000`. Existing users who require unlimited shots can pass `max_shots=None`. The previous `diagonalisation_method` spelling remains available as a deprecated compatibility alias.

## What was changed?

- Added `shots` to request an explicit total shot count.
  It takes precedence over `precision` and is distributed across measurement groups proportionally to their estimated variances using largest-remainder rounding.
- Added `max_shots` as a hard upper bound for both precision-based and explicit shot allocations.
  It must be a positive integer or `None`; violations are detected before backend or sampler execution.
- Kept the existing `precision` parameter and documented its dependence on both inverse-square precision scaling and Hamiltonian coefficient magnitudes.
- Renamed the preferred public parameter to `diagonalization_method`.
  The former `diagonalisation_method` spelling remains supported with a deprecation warning, and passing both spellings raises `ValueError`.
- Added a shared backend-independent measurement planner for Hamiltonian normalization, grouping, basis-transformed operators, variance estimates, and shot allocation.
- Reused the planner in both static backend execution and Jasp/JAX execution.
- Added type annotations and clarified static `float` versus Jasp zero-dimensional `jax.Array` return behavior.
- Updated VQE, QITE, examples, regression tests, documentation, and the developer changelog.
- Cleaned up Ruff findings introduced by the new measurement code.

## How was it tested?

| Test-ID | Status |
|---------|--------|
| T-001 | ✅ `pytest tests/operators_tests/test_expectation_value.py tests/operators_tests/test_measurement_plan.py` — 13 passed |
| T-002 | ✅ `pytest tests/operators_tests/test_qubit_operator.py` — 1 passed |
| T-003 | ✅ Targeted Ruff checks for the new measurement modules — all checks passed |
| T-004 | ✅ Diagnostics and `git diff --check` — clean |

**Input validation tests:**

| Test | Validation case | Input | Expected result | Validation timing |
|---|---|---|---|---|
| `test_expectation_value_rejects_invalid_parameters_at_construction` | Zero precision | `precision=0` | Raises `ValueError` | Construction |
| Same parameterized test | Negative precision | `precision=-0.1` | Raises `ValueError` | Construction |
| Same parameterized test | Non-finite precision | `precision=NaN` | Raises `ValueError` | Construction |
| Same parameterized test | Invalid shots | `shots=0` | Raises `ValueError` | Construction |
| Same parameterized test | Invalid shot limit | `max_shots=0` | Raises `ValueError` | Construction |
| Same parameterized test | Unknown diagonalization method | `diagonalization_method="unknown"` | Raises `ValueError` | Construction |
| `test_expectation_value_validates_empty_hamiltonian_parameters` | Invalid shots for empty Hamiltonian | `0 * X(0)`, `shots=0` | Raises `ValueError` | Construction |
| `test_expectation_value_shots_take_precedence_over_invalid_precision` | Explicit shots override precision | `shots=10`, `precision=0` | Accepted successfully | Construction |
| `test_expectation_value_default_max_shots` | Estimated shot budget exceeds default limit | `precision=0.001` | Raises `ValueError` mentioning `max_shots=100000` | Evaluation |
| `test_expectation_value_rejects_both_diagonalization_spellings` | Conflicting argument aliases | Both diagonalization spellings supplied | Raises `ValueError` | Construction |
| `test_expectation_value_accepts_deprecated_diagonalisation_alias` | Deprecated alias compatibility | `diagonalisation_method="commuting_qw"` | Emits `QrispDeprecationWarning` | Evaluation |

The focused validation suite currently passes: **20 tests passed**.

## Screenshots / Output

Not applicable.

## Checklist

- [x] My code follows the project's coding standards
- [x] I have performed a self-review
- [x] I have added/updated tests
- [ ] All tests pass locally and in CI
- [x] I have updated the documentation
- [x] I have added a changelog entry to `changelog-dev.rst`
- [x] Breaking changes are documented with migration path

## Reviewer Notes

Please review the default `max_shots=100_000` behavior and confirm that the explicit `max_shots=None` escape hatch is the intended migration path.

Please also verify the shot-allocation semantics:

- `shots` overrides `precision`.
- Precision-based allocation uses operator-derived variance weights.
- `max_shots` applies to both allocation modes and is checked before execution.
- Invalid diagonalization-method combinations are rejected, while the deprecated British spelling remains compatible.

- The `max_shots` cap **will require updating Tutorial and Examples** in the documentation from, e.g.:

```
vqe = heisenberg_problem(G,1,1)
vqe.set_callback()
energy = vqe.run(QuantumVariable(G.number_of_nodes()),depth=2,max_iter=50)
```

to

```
vqe = heisenberg_problem(G,1,1)
vqe.set_callback()
energy = vqe.run(QuantumVariable(G.number_of_nodes()),depth=2,max_iter=50, mes_kwargs={"max_shots": None})
```

to deactivate the `max_shots` bound or set an appropriate value. 
This hardcoded and somewhat arbitrary `max_shots` bound will therefore make the code look more clunky.
We could also remove `max_shots` and only add the optional `shots` parameter. But, then users could still run into executing prohibitively high shot numbers. Or we to use `shots=5000` as default and add `atol` and `rtol` for absolute and relative precision (that override `shots`). There are many possibilities.

@bastibock @positr0nium what do you think?

## Relative precision

Replacing absolute precision entirely would be risky. Relative precision based on the Hamiltonian scale is valuable, but the two modes serve different purposes.

For a scale $S(H)$, relative precision means:

$\epsilon_{\mathrm{absolute}}=\epsilon_{\mathrm{relative}} S(H).$

**Relative precision is useful when:**

- Hamiltonians are routinely rescaled.
- Results from differently scaled Hamiltonians are compared.
- The desired accuracy is proportional to the operator magnitude.
- You want shot counts to be approximately invariant under $H \mapsto cH$.

**Absolute precision is important when:**

- The user needs a fixed physical tolerance, such as chemical accuracy.
- VQE convergence depends on resolving small energy differences.
- The Hamiltonian has a large constant identity offset.
- The expectation is small because of cancellation, but that small value is meaningful.
- Units have physical meaning and the error must be bounded in those units.

For example, if

$H = 1000I + 0.001Z,$

a norm including the identity term would make a relative tolerance almost useless for estimating the physically interesting $Z$ contribution. Since identity terms do not contribute sampling variance, the scale definition needs care.

My recommendation is:

- Keep absolute `precision` as the default.
- Add an opt-in `relative_precision`.
- Make the two options mutually exclusive.
- Document exactly which coefficient/operator scale is used.
- Do not silently change the meaning of the existing `precision` argument.

That provides scale-invariant behavior for users who want it without breaking fixed-error use cases.

### CODE REVIEW

**Comment by renezander90** [comment]
Also tagging @purva-thakre : I used this PR to deprecate the `diagonalisation_method` (BE) parameter in favour of `diagonalization_method` (AE).
